src/XrdHttp/XrdHttpProtocol.cc

```diff
@@ -1759,10 +1759,17 @@ void XrdHttpProtocol::Cleanup() {
 
   if (ssl) {
 
-
-    if (SSL_shutdown(ssl) != 1) {
-      TRACE(ALL, " SSL_shutdown failed");
-      ERR_print_errors(sslbio_err);
+    int ret = SSL_shutdown(ssl);
+    if (ret != 1) {
+        if(ret == 0) {
+            //https://www.openssl.org/docs/man1.0.2/man3/SSL_shutdown.html
+            //Call again SSL_shutdown
+            ret = SSL_shutdown(ssl);
```

Comment:
So, tell me why we need a bidirectional shutdown here. From the same documentation "If a unidirectional shutdown is enough (the underlying connection shall be closed anyway), this first call to SSL_shutdown() is sufficient." I think the actual problem was that the original code insisted that the return value be one when in fact 1 or 0 was good enough and only -1 should have reported an error. However, a zero return should clear the error stack as a misleading error may have been placed on the stack. Agree?

Reply:
> OK, this will work but it does slow the connection turnaround because it tries to do a bidirectional shutdown which I don't think is needed. So, while not technically wrong there is a possible improvement and I leave it up to you to decide which way to go.

The documentation says that bidirectional shutdown is necessary if the connection is going to be reused. If the connection will be closed anyway, then the unidirectional shutdown is enough. Therefore, if we are going to reuse the connection in question here, we should do the complete shutdown, as done by @ccaffy here. However, I think from the above, before calling any shutdown function, we must check that there are no pre-existing errors on the connection.

Reply:
Instead of calling `ERR_print_errors(bio*)` can't we just clear the thread local queue, and if we really need just peek on the last error code. Depending on how openssl implements its thread local vs global queues it might give us a slight edge over trying to use ERR_print_errors ?

https://www.openssl.org/docs/man1.0.2/man3/ERR_remove_thread_state.html & 
https://www.openssl.org/docs/man1.0.2/man3/ERR_peek_last_error.html

Reply:
I was thinking about calling `SSL_get_error`, as recommended in the docs.

Reply:
> I was thinking about calling `SSL_get_error`, as recommended in the docs.

The current docs only mention this function as a switch statement for a few functions (not including ssl_shutdown), not sure whether the doc. is reflective of reality tho :/

"SSL_get_error() returns a result code (suitable for the C "switch" statement) for a preceding call to SSL_connect(), SSL_accept(), SSL_do_handshake(), SSL_read(), SSL_peek(), or SSL_write() on ssl."

Reply:
What does "reuse connection" really mean? The socket stays open and connected and we start a new session on the existing socket?  When Recycle() gets called the socket will definitely be closed unless it was marked as a shared socket which doesn't appear to ever happen.

Reply:
> I was thinking about calling `SSL_get_error`, as recommended in the docs.

It looks like I need an error code as second argument, which I do not have right now. There's only one place where `SSL_get_error()` is called: https://github.com/xrootd/xrootd/blob/master/src/XrdHttp/XrdHttpProtocol.cc#L531

> What does "reuse connection" really mean? The socket stays open and connected and we start a new session on the existing socket? When Recycle() gets called the socket will definitely be closed unless it was marked as a shared socket which doesn't appear to ever happen.

Alright so thanks to that architecture, there is no need to call this SSL_Shutdown() twice as the connection will never be re-used. Thanks!

Reply:
OK. Could you make sure the comment reflects this as I can envision this conversation happening again several years from now.

Reply:
@abh3 done and documented :)